### DISCUSSION TITLE
Best Practice for Java-based Apps

### DISCUSSION THREAD

**Discussion by wcarson** [microsoft/winget-pkgs#11247] [Q&A]
What is the best practice for Java-based apps that have a dependency on a minimum Java version being available on the system? For example, if I'm writing a (v1.0.0) manifest for an Eclipse-based app which requires a minimum of Java 11 on the target system. How is that dependency to be reflected in the manifest if at all?

**Comment by denelon** [comment]
We've been talking about complex dependencies like Java. We're looking at a model where we display something informational like "This package depends on ```<<ExternalDependencies>>```. If your system has the dependencies would you like to continue (y/n)?"

We also had an ask about supporting URLs for people to research their choices. Optionally, we could also link to any packages in the repository that meet those dependencies like "Oracle.JavaRuntimeEnvironment" or "AdoptOpenJDK.OpenJDK" etc.

The client doesn't support dependencies https://github.com/microsoft/winget-cli/issues/163 yet, nor does our validation environment. This work will need to be completed before we can properly articulate dependencies in a manifest.

**Comment by brunoborges** [comment]
If your application requires Java 11+, you should ideally be shipping a Java runtime bundled with your application, instead of asking for it to be available in the environment.

Please look into the `jlink` command in the JDK 11 and how to create such Java runtime.

In Java 16, there is also a new command line `jpackage` in the JDK to create package installers for Windows, macOS, and Linux. I'd suggest you start there.

**Comment by wcarson** [comment]
@brunoborges - I agree. Unfortunately, this isn't an app under my control, I'm just creating a winget manifest for it.

**Comment by AwesomeKalin** [comment]
What is the app?

**Comment by wcarson** [comment]
@AwesomeKalin - Apache Directory Studio

**Comment by brunoborges** [comment]
@wcarson if this application is not under your control, I'd suggest to post a message to their mailing list and start a discussion there, so they can address the problem regarding Java runtime bundling.
